crates/ruff_cli/src/commands/format.rs

```diff
@@ -145,24 +146,111 @@ fn format_path(
     source_type: PySourceType,
     mode: FormatMode,
 ) -> Result<FormatCommandResult, FormatCommandError> {
-    let unformatted = std::fs::read_to_string(path)
-        .map_err(|err| FormatCommandError::Read(Some(path.to_path_buf()), err))?;
-
-    let options = settings.to_format_options(source_type, &unformatted);
-    debug!("Formatting {} with {:?}", path.display(), options);
-
-    let formatted = format_module_source(&unformatted, options)
-        .map_err(|err| FormatCommandError::FormatModule(Some(path.to_path_buf()), err))?;
+    // Extract the sources from the file.
+    let source_kind = match SourceKind::from_path(path, source_type) {
+        Ok(Some(source_kind)) => source_kind,
+        Ok(None) => return Ok(FormatCommandResult::Unchanged),
+        Err(err) => {
+            return Err(FormatCommandError::Read(Some(path.to_path_buf()), err));
+        }
+    };
 
-    let formatted = formatted.as_code();
-    if formatted.len() == unformatted.len() && formatted == unformatted {
-        Ok(FormatCommandResult::Unchanged)
-    } else {
+    // Format the source.
+    if let Some(source_kind) = format_source(&source_kind, source_type, Some(path), settings)? {
         if mode.is_write() {
-            std::fs::write(path, formatted.as_bytes())
+            let mut writer = File::create(path)
+                .map_err(|err| FormatCommandError::Write(Some(path.to_path_buf()), err.into()))?;
+            source_kind
+                .write(&mut writer)
                 .map_err(|err| FormatCommandError::Write(Some(path.to_path_buf()), err))?;
         }
         Ok(FormatCommandResult::Formatted)
+    } else {
+        Ok(FormatCommandResult::Unchanged)
+    }
+}
+
+/// Format a [`SourceKind`], returning the transformed [`SourceKind`], or `None` if the source was
+/// unchanged.
+pub(crate) fn format_source(
+    source_kind: &SourceKind,
+    source_type: PySourceType,
+    path: Option<&Path>,
+    settings: &FormatterSettings,
+) -> Result<Option<SourceKind>, FormatCommandError> {
+    match source_kind {
+        SourceKind::Python(unformatted) => {
+            let options = settings.to_format_options(source_type, unformatted);
+
+            let formatted = format_module_source(unformatted, options)
+                .map_err(|err| FormatCommandError::Format(path.map(Path::to_path_buf), err))?;
+
+            let formatted = formatted.into_code();
+            if formatted.len() == unformatted.len() && formatted == *unformatted {
+                Ok(None)
+            } else {
+                Ok(Some(SourceKind::Python(formatted)))
+            }
+        }
+        SourceKind::IpyNotebook(notebook) => {
+            if !notebook.is_python_notebook() {
+                return Ok(None);
+            }
+
+            let mut output = String::with_capacity(notebook.source_code().len());
```

Comment:
We can avoid the allocation here by only reserving once we see the first formatted cell